fdbserver/ptxn/TLogServer.actor.cpp

```diff
@@ -1363,21 +1367,153 @@ Reference<LogGenerationData> findLogData(
+	// this might be an off-by-one error from other places.
+	if (req.beginVersion <= logData->persistentDataDurableVersion) {
+		// Just in case the durable version changes while we are waiting for the read, we grab this data from memory. We
+		// may or may not actually send it depending on whether we get enough data from disk. SOMEDAY: Only do this if
+		// an initial attempt to read from disk results in insufficient data and the required data is no longer in
+		// memory SOMEDAY: Should we only send part of the messages we collected, to actually limit the size of the
+		// result?
+
+		if (reqOnlySpilled) {
+			endVersion = logData->persistentDataDurableVersion + 1;
+		} else {
+			peekMessagesFromMemory(req, values, beginVersion, endVersion, logData);
+		}
+
+		if (logData->shouldSpillByValue(req.storageTeamID)) {
+			RangeResult kvs = wait(tlogGroup->second->persistentData->readRange(
+			    KeyRangeRef(persistStorageTeamMessagesKey(logData->logId, req.storageTeamID, req.beginVersion),
+			                persistStorageTeamMessagesKey(
+			                    logData->logId, req.storageTeamID, logData->persistentDataDurableVersion + 1)),
+			    TLOG_PEEK_REQUEST_REPLY_SIZE_CRITERIA,
+			    TLOG_PEEK_REQUEST_REPLY_SIZE_CRITERIA));
+			bool first = true;
+			for (auto& kv : kvs) {
+				auto ver = decodeStorageTeamMessagesKey(kv.key);
+				// try decode kv.value here, who is encoded by proxy
+				versionsFromDisk.insert(ver);
+				serializer.writeSerializedVersionSection(
+				    kv.value); // if comment this line, then can read the in-memory data, so problem is here
+				if (first) {
+					beginVersion = ver;
+					first = false;
+				}
+			}
+			if (kvs.expectedSize() >=
+			    TLOG_PEEK_REQUEST_REPLY_SIZE_CRITERIA) { // if enough from disk, not reading from memory
+				endVersion = decodeStorageTeamMessagesKey(kvs.end()[-1].key) + 1;
+				onlySpilled = true;
+			} else {
+				serializeMemoryData(values, serializer, versionsFromDisk);
+			}
+		} else {
+			// TODO: make it work for spill by reference,
+			// now writing(and only writing) those in-memory data without considering on-disk data
+			serializeMemoryData(values, serializer);
+		}
+	} else {
+		if (reqOnlySpilled) {
+			endVersion = logData->persistentDataDurableVersion + 1;
+		} else {
+			peekMessagesFromMemory(req, values, beginVersion, endVersion, logData);
+			serializeMemoryData(values, serializer);
+		}
+		//TraceEvent("TLogPeekResults", self->dbgid).detail("ForAddress", replyPromise.getEndpoint().getPrimaryAddress()).detail("MessageBytes", messages.getLength()).detail("NextEpoch", next_pos.epoch).detail("NextSeq", next_pos.sequence).detail("NowSeq", self->sequence.getNextSequence());
+	}
+
+	TLogPeekReply reply;
+	reply.maxKnownVersion = logData->version.get();
+	reply.minKnownCommittedVersion = logData->minKnownCommittedVersion;
+	reply.data = StringRef(reply.arena, serializer.getSerialized());
```

Comment:
`serializer.getSerialized()` returns a `Standalone<StringRef>`. Even if you do this, the returned `arena` is not copied, so the memory is could be gone after this line. The correct way is:
```
auto replyData = serializer.getSerialized(); // this holds the returned arena
reply.data = replyData;
```

Reply:
You mean currently i have a standalone return value but there is no reference to it,  so the arena of this standalone might be lost, Thus we need to have a reference?

Reply:
That's correct.

Reply:
cool,i am doing as below, as I think we need to use the arena in the reply

```
auto replyData = serializer.getSerialized(); // this holds the returned arena
reply.data = StringRef(reply.arena, replyData)
```

Reply:
correct me if I am wrong

Reply:
Doing `reply.data = StringRef(reply.arena, replyData)` is correct, but introduces an extra `memcpy` that we want to avoid.

Reply:
oh, so without explicitly saying we want to use `reply.arena`, we can reference the existing arena? will fix

Reply:
Actually if I use `reply.data = replyData;` then I start to see the error below, can you please double check?

```
trace.0.0.0.0.0.1645680553.o8XOXR.0.1.xml:<Event Severity="40" ErrorKind="BugDetected" Time="88.114987" DateTime="2022-02-24T05:29:15Z" Type="InternalError" Machine="[abcd::3:4:3:6]:1" ID="0000000000000000" Error="internal_error" ErrorDescription="An internal error occurred" ErrorCode="4100" FailedAssertion="e &lt;= end" File="/root/src/foundationdb/flow/serialize.h" Line="624" ThreadID="4134975440075990537" Backtrace="addr2line -e fdbserver.debug -p -C -f -i 0x34e3f6c 0x34e31b0 0x34e3591 0x343d10c 0x1a247bd 0x1a24d4a 0x1a25a50 0x1a260fd 0x1a2626a 0x1a26d22 0x1a315c3 0x1a30ed9 0x1a36b7d 0x1a36fd8 0x1a384e4 0x1a3875e 0x1273cd8 0xfc59a0 0x3381f92 0x33abdd0 0x33820ad 0x33abb5f 0xfc59a0 0x3481633 0x33abfe2 0xa51bef 0x7fe4a17c6555" LogGroup="default" Roles="TL,TS" />
```

Reply:
Can you try adding one line:
```
reply.arena.dependsOn(replyData.arena());
```

Reply:
it works! So I guess as soon as we quit this method, the arena of this `replyData` will be cleaned, and thus we either need to memcpy to `reply.arena`(previously), or we need to have this explicit reference for `replyData.arena`(now)?

Reply:
Yes. Either `memcpy` or explicitly hold the reference.

Reply:
good to learn this! I have fixed it, joshua seems fine 20220224-162602-haofu-9c1af27a48dec20b